### PR TITLE
Add AlphaMask2d phase

### PR DESCRIPTION
**Pull request by IceSentry**
# Objective

- Bevy now supports an opaque phase for mesh2d, but it's very common for 2d textures to have a transparent alpha channel.

## Solution

- Add an alpha mask phase identical to the one in 3d. It will do the alpha masking in the shader before drawing the mesh.
- Uses the BinnedRenderPhase
- Since it's an opaque draw it also correctly writes to depth

## Testing

- Tested the mes2d_alpha_mode example and the bevymark example with alpha mask mode enabled

---

## Showcase

image

The white logo on the right is rendered with alpha mask enabled.

Running the bevymark example I can get 65fps for 120k mesh2d all using alpha mask.

## Notes

This is one more step for mesh2d improvements https://github.com/bevyengine/bevy/issues/13265

### CODE REVIEW

File: crates/bevy_sprite/src/mesh2d/material.rs [hunk 1]
```diff
@@ -149,6 +149,7 @@ pub enum AlphaMode2d {
     /// Base color alpha values are overridden to be fully opaque (1.0).
     #[default]
     Opaque,
+    Mask(f32),
```
Comment (tychedelia) [inline crates/bevy_sprite/src/mesh2d/material.rs:152]: nit: missing doc for users

**Review by tychedelia — APPROVED** [review]
this is very didactic to see this built up in layers. looks good, thanks!

File: crates/bevy_core_pipeline/src/core_2d/mod.rs [hunk 2]
```diff
@@ -207,6 +209,94 @@ impl CachedRenderPipelinePhaseItem for Opaque2d {
...
+    /// The function used to draw.
+    pub draw_function: DrawFunctionId,
+    /// The asset that this phase item is associated with.
+    ///
+    /// Normally, this is the ID of the mesh, but for non-mesh items it might be
+    /// the ID of another type of asset.
+    pub asset_id: UntypedAssetId,
+    /// The ID of a bind group specific to the material.
+    ///
+    /// In the case of PBR, this is the `MaterialBindGroupId`.
```
Comment (kristoff3r) [inline crates/bevy_core_pipeline/src/core_2d/mod.rs:240]: I guess this isn't relevant for 2D? Maybe there is a better example in this case.

File: crates/bevy_sprite/src/mesh2d/color_material.rs [hunk 3]
```diff
@@ -97,17 +97,30 @@ impl From<Handle<Image>> for ColorMaterial {
 bitflags::bitflags! {
     #[repr(transparent)]
     pub struct ColorMaterialFlags: u32 {
-        const TEXTURE       = 1 << 0;
-        const NONE          = 0;
-        const UNINITIALIZED = 0xFFFF;
+        const TEXTURE                    = 1 << 0;
+        // Bitmask reserving bits for the [`AlphaMode2d`]
+        // Values are just sequential values bitshifted into
+        // the bitmask, and can range from 0 to 3.
```
Comment (kristoff3r) [inline crates/bevy_sprite/src/mesh2d/color_material.rs:103]:
```suggestion
        /// Bitmask reserving bits for the [`AlphaMode2d`]
        /// Values are just sequential values bitshifted into
        /// the bitmask, and can range from 0 to 3.
```

File: crates/bevy_sprite/src/mesh2d/color_material.wgsl [hunk 4]
```diff
@@ -11,8 +11,14 @@ struct ColorMaterial {
...
     flags: u32,
+    alpha_cutoff: f32,
 };
-const COLOR_MATERIAL_FLAGS_TEXTURE_BIT: u32 = 1u;
+
+const COLOR_MATERIAL_FLAGS_TEXTURE_BIT: u32              = 1u;
+const COLOR_MATERIAL_FLAGS_ALPHA_MODE_RESERVED_BITS: u32 = 3221225472u; // (0b11u32 << 30)
+const COLOR_MATERIAL_FLAGS_ALPHA_MODE_OPAQUE: u32        = 0u;          // (0u32 << 30)
+const COLOR_MATERIAL_FLAGS_ALPHA_MODE_MASK: u32          = 1073741824u; // (1u32 << 30)
+const COLOR_MATERIAL_FLAGS_ALPHA_MODE_BLEND: u32         = 2147483648u; // (2u32 << 30)
```
Comment (kristoff3r) [inline crates/bevy_sprite/src/mesh2d/color_material.wgsl:21]: I would just write the expressions, constants are guaranteed to be evaluated at compile time (see https://doc.rust-lang.org/reference/const_eval.html#constant-expressions)

File: crates/bevy_sprite/src/mesh2d/color_material.wgsl [hunk 5]
```diff
@@ -23,14 +29,40 @@ fn fragment(
...
 }
+
+fn alpha_discard(material: ColorMaterial, output_color: vec4<f32>) -> vec4<f32> {
+    var color = output_color;
+    let alpha_mode = material.flags & COLOR_MATERIAL_FLAGS_ALPHA_MODE_RESERVED_BITS;
+    if alpha_mode == COLOR_MATERIAL_FLAGS_ALPHA_MODE_OPAQUE {
+        // NOTE: If rendering as opaque, alpha should be ignored so set to 1.0
+        color.a = 1.0;
+    }
+    // TODO MAY_DISCARD
```
Comment (kristoff3r) [inline crates/bevy_sprite/src/mesh2d/color_material.wgsl:56]: Should either be fixed or have some context, like "add MAY_DISCARD ifdef" (I had to search to figure out what it meant).

**Review by kristoff3r — APPROVED** [review]
LGTM apart from some nits. I also tested the mesh2d_alpha_mode example on wasm.

File: crates/bevy_core_pipeline/src/core_2d/mod.rs (hunk 2 above)
Comment (IceSentry) [reply crates/bevy_core_pipeline/src/core_2d/mod.rs:240]: Oh, good catch, I just copied the docs verbatim from 3d 😅

File: crates/bevy_sprite/src/mesh2d/color_material.wgsl (hunk 5 above)
Comment (IceSentry) [reply crates/bevy_sprite/src/mesh2d/color_material.wgsl:56]: Oh, right, I completely forgot that wasn't implemented, I'll implement it before it gets merged.

**Comment by IceSentry** [comment]
I added the blocked label because I need to implement the `MAY_DISCARD` shader def

File: crates/bevy_sprite/src/mesh2d/color_material.wgsl (hunk 4 above)
Comment (IceSentry) [reply crates/bevy_sprite/src/mesh2d/color_material.wgsl:21]: It's not supported in wgsl unfortunately. We should maybe look into using shader_defs for those though.

Comment (kristoff3r) [reply crates/bevy_sprite/src/mesh2d/color_material.wgsl:21]: Oh I didn't even notice this snippet was in a wgsl file, the syntax is too similar :sweat_smile:  Sorry about that.

**Pull request merged**